src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs

```diff
@@ -589,57 +589,57 @@ public ResultWrapper<IEnumerable<FilterLog>> eth_getFilterLogs(UInt256 filterId)
 
     public ResultWrapper<IEnumerable<FilterLog>> eth_getLogs(Filter filter)
     {
+        BlockParameter fromBlock = filter.FromBlock;
+        BlockParameter toBlock = filter.ToBlock;
+
         // because of lazy evaluation of enumerable, we need to do the validation here first
         CancellationTokenSource timeout = BuildTimeoutCancellationTokenSource();
         CancellationToken cancellationToken = timeout.Token;
 
+        if (!TryFindOrUseLatest(_blockFinder, ref toBlock, out SearchResult<BlockHeader> toBlockResult, out long sourceToBlockNumber))
+        {
+            timeout.Dispose();
+            return FailWithNoHeadersSyncedYet(toBlockResult);
+        }
+
+        cancellationToken.ThrowIfCancellationRequested();
+
         SearchResult<BlockHeader> fromBlockResult;
-        SearchResult<BlockHeader> toBlockResult;
+        long sourceFromBlockNumber;
 
-        if (filter.FromBlock == filter.ToBlock)
+        if (fromBlock == toBlock)
         {
-            fromBlockResult = toBlockResult = _blockFinder.SearchForHeader(filter.ToBlock);
+            fromBlockResult = toBlockResult;
+            sourceFromBlockNumber = sourceToBlockNumber;
         }
-        else
+        else if (!TryFindOrUseLatest(_blockFinder, ref fromBlock, out fromBlockResult, out sourceFromBlockNumber))
         {
-            toBlockResult = _blockFinder.SearchForHeader(filter.ToBlock);
-
-            if (toBlockResult.IsError)
-            {
-                timeout.Dispose();
-                return GetFailureResult<IEnumerable<FilterLog>, BlockHeader>(toBlockResult, _ethSyncingInfo.SyncMode.HaveNotSyncedHeadersYet());
-            }
-
-            cancellationToken.ThrowIfCancellationRequested();
-            fromBlockResult = _blockFinder.SearchForHeader(filter.FromBlock);
+            timeout.Dispose();
+            return FailWithNoHeadersSyncedYet(fromBlockResult);
         }
 
-        if (fromBlockResult.IsError)
+        if (sourceFromBlockNumber > sourceToBlockNumber)
         {
             timeout.Dispose();
-            return GetFailureResult<IEnumerable<FilterLog>, BlockHeader>(fromBlockResult, _ethSyncingInfo.SyncMode.HaveNotSyncedHeadersYet());
+            return ResultWrapper<IEnumerable<FilterLog>>.Fail($"From block {sourceFromBlockNumber} is later than to block {sourceToBlockNumber}.", ErrorCodes.InvalidParams);
         }
 
-        cancellationToken.ThrowIfCancellationRequested();
-
-        BlockHeader fromBlock = fromBlockResult.Object!;
-        BlockHeader toBlock = toBlockResult.Object!;
-
-        long fromBlockNumber = fromBlock.Number;
-        long toBlockNumber = toBlock.Number;
-
-        if (fromBlockNumber > toBlockNumber && toBlockNumber != 0)
+        if (_blockFinder.Head?.Number is not null && sourceFromBlockNumber > _blockFinder.Head.Number)
```

Comment:
I see that  `&& toBlockNumber != 0` condition was removed, is this case handled properly?